src/app/space/settings/areas/areas.component.ts

```diff
@@ -18,69 +30,251 @@ export class AreasComponent implements OnInit, OnDestroy {
+    this.subscriptions.forEach(sub => {
+      sub.unsubscribe();
+    });
   }
 
-  openModal() {
+  openModal(): void {
     this.modal.show();
   }
 
-  addChildArea(id: string) {
+  addChildArea(id: string): void {
     this.selectedAreaId = this.defaultArea;
     if (id) {
       this.selectedAreaId = id;
     }
     this.openModal();
   }
 
-  itemPath(item: AreaAttributes) {
-    // remove slash from start of string
-    let parentPath = item.parent_path_resolved.slice(1, item.parent_path_resolved.length);
-    if (parentPath === '') {
-      return item.name;
+  addArea(area: Area): void {
+    this.allAreas.push(area);
+    this.showTree = false;
+
+    // Hide the tree list to rerender existing toggles in expanded state
+    setTimeout(() => {
+      // Reapply sort and filter, if any
+      this.allAreas.sort((area1: Area, area2: Area) => this.compare(area1, area2));
+      this.applyFilters(this.appliedFilters);
+      this.showTree = true;
+    }, 10);
+  }
+
+  // Filter
+
+  applyFilters(filters: Filter[]): void {
+    this.appliedFilters = filters;
+    this.filteredAreas = [];
+    if (filters && filters.length > 0) {
+      this.allAreas.forEach((area) => {
+        if (this.matchesFilters(area, filters)) {
+          this.filteredAreas.push(cloneDeep(area));
+        }
+      });
+      this.treeAreas = this.buildFilteredTree(this.filteredAreas);
+    } else {
+      this.treeAreas = this.buildTree(this.allAreas);
+    }
+    this.resultsCount = this.filteredAreas.length;
+  }
+
+  filterChange($event: FilterEvent): void {
+    this.applyFilters($event.appliedFilters);
+  }
+
+  matchesFilter(area: Area, filter: Filter): boolean {
+    let match = true;
+
+    if (filter.field.id === 'area') {
+      match = area.attributes.name.match(filter.value) !== null;
     }
-    return parentPath + '/' + item.name;
+    return match;
   }
 
-  addArea(area: Area) {
-    this.areas.push(area);
+  matchesFilters(area: Area, filters: Filter[]): boolean {
+    let matches = true;
+
+    filters.forEach((filter) => {
+      if (!this.matchesFilter(area, filter)) {
+        matches = false;
+        return false;
+      }
+    });
+    return matches;
+  }
+
+  // Sort
+
+  compare(area1: Area, area2: Area): number {
+    var compValue = 0;
+    if (this.currentSortField === undefined || this.currentSortField.id === 'area') {
+      compValue = area1.attributes.name.localeCompare(area2.attributes.name);
+    }
+    if (!this.isAscendingSort) {
+      compValue = compValue * -1;
+    }
+    return compValue;
+  }
+
+  sortChange($event: SortEvent): void {
+    this.currentSortField = $event.field;
+    this.isAscendingSort = $event.isAscending;
+    this.allAreas.sort((area1: Area, area2: Area) => this.compare(area1, area2));
+    this.applyFilters(this.appliedFilters); // Reapply filters, if any
+  }
+
+  // Transform flat array into tree list
+
+  private buildTree(elements: any[], tree = []): any[] {
```

Comment:
I don't agree with the use of `any` in order to add properties that don't exist to Typed objects. I would rather have the component define a new type that contains an Area along with it's extra tree information. It then becomes obvious to any developer unfamiliar with the code what the object contains and what things are used for.

Reply:
I'll extend the interface